### PR TITLE
More bug fixes for Service Discovery

### PR DESCRIPTION
**Pull request by alena1108**
1) SD external_links:  follow container:link semantic in config

https://github.com/rancherio/rancher/issues/631

2) When service is removed, cleanup service maps referencing service-to-remove as:

* consuming service
* consumed service

https://github.com/rancherio/rancher/issues/638
https://github.com/rancherio/rancher/issues/643

3) Don't allow null value for scale parameter:

https://github.com/rancherio/rancher/issues/634

4) LB AppCookieStickinessPolicy: marked cookie/maxLength/timeout as required in API as those are required by haproxy

### CODE REVIEW

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java [hunk 1]
```diff
@@ -273,7 +273,7 @@ private void populateExternalLinksForService(Service service, Map<String, Object
                         null);
                 String instanceName = getInstanceName(instance);
                 if (instanceName != null) {
-                    instanceLinksWithNames.put(linkName, instanceName);
+                    instanceLinksWithNames.put(instanceName, linkName);
```
Comment (sonchang) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java]: darn.  sorry I missed this the first time around

**Comment by sonchang** [comment]
LGTM

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java (hunk 1 above)
Comment (alena1108) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java]: no worries, totally my bad, should have paid more attention

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/dao/impl/ServiceConsumeMapDaoImpl.java [hunk 2]
```diff
@@ -66,4 +66,15 @@ public ServiceConsumeMap findNonRemovedMap(long serviceId, long consumedServiceI
...
     }
 
+    @Override
+    public List<? extends ServiceConsumeMap> findConsumingMapsToRemove(long serviceId) {
+        return create()
+                .selectFrom(SERVICE_CONSUME_MAP)
+                .where(
+                        SERVICE_CONSUME_MAP.CONSUMED_SERVICE_ID.eq(serviceId)
+                                .and((SERVICE_CONSUME_MAP.REMOVED.isNull().
+                                        or(SERVICE_CONSUME_MAP.STATE.eq(CommonStatesConstants.REMOVING))))).
```
Comment (cjellick) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/dao/impl/ServiceConsumeMapDaoImpl.java]: Want to make sure I understand this query, you want to get maps "where removed is null or state is removing."
Is that right?
Seems to me like it should be "where removed is null and state is not removing"?

Comment (alena1108) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/dao/impl/ServiceConsumeMapDaoImpl.java]: Here I'm finding the maps To remove; so I specifically looking for maps with removed flag null, or state=Removing (records against which Remove process was called once). I follow the logic from genericMapDao here

**Comment by cjellick** [comment]
LGTM

**Pull request merged**